infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/UserPreferencesConfigurator.java

```diff
@@ -0,0 +1,133 @@
+    this.namespaceFactory = namespaceFactory;
+    this.clientFactory = clientFactory;
+    this.userManager = userManager;
+    this.preferenceManager = preferenceManager;
+  }
+
+  @Override
+  public void configure(NamespaceResolutionContext namespaceResolutionContext)
+      throws InfrastructureException {
+    Secret userPreferencesSecret = preparePreferencesSecret(namespaceResolutionContext);
+    String namespace = namespaceFactory.evaluateNamespaceName(namespaceResolutionContext);
+
+    try {
+      clientFactory
+          .create()
+          .secrets()
+          .inNamespace(namespace)
+          .createOrReplace(userPreferencesSecret);
+    } catch (KubernetesClientException e) {
+      throw new InfrastructureException(
+          "Error occurred while trying to create user preferences secret.", e);
+    }
+  }
+
+  private Secret preparePreferencesSecret(NamespaceResolutionContext namespaceResolutionContext)
+      throws InfrastructureException {
+    Base64.Encoder enc = Base64.getEncoder();
+    User user;
+    Map<String, String> preferences;
+
+    try {
+      user = userManager.getById(namespaceResolutionContext.getUserId());
+      preferences = preferenceManager.find(user.getId());
+    } catch (NotFoundException | ServerException e) {
+      throw new InfrastructureException(
+          String.format(
+              "Preferences of user with id:%s cannot be retrieved.",
+              namespaceResolutionContext.getUserId()),
+          e);
+    }
+
+    if (preferences == null || preferences.isEmpty()) {
+      throw new InfrastructureException(
+          String.format(
+              "Preferences of user with id:%s are empty. Cannot create user preferences secrets.",
+              namespaceResolutionContext.getUserId()));
+    }
+
+    Map<String, String> preferencesEncoded = new HashMap<>();
+    preferences.forEach(
+        (key, value) ->
+            preferencesEncoded.put(
+                normalizePreferenceName(key), enc.encodeToString(value.getBytes())));
+    return new SecretBuilder()
+        .addToData(preferencesEncoded)
+        .withNewMetadata()
+        .withName(USER_PREFERENCES_SECRET_NAME)
+        .addToLabels(DEV_WORKSPACE_MOUNT_LABEL, "true")
+        .addToAnnotations(DEV_WORKSPACE_MOUNT_AS_ANNOTATION, "file")
+        .addToAnnotations(DEV_WORKSPACE_MOUNT_PATH_ANNOTATION, USER_PREFERENCES_SECRET_MOUNT_PATH)
+        .endMetadata()
+        .build();
+  }
+
+  /**
+   * Some preferences names are not compatible with k8s restrictions on key field in secret. The
+   * keys of data must consist of alphanumeric characters, -, _ or . This method replaces illegal
+   * characters with -
+   *
+   * @param name original preference name
+   * @return k8s compatible preference name used as a key field in Secret
+   */
+  @VisibleForTesting
+  String normalizePreferenceName(String name) {
+    return name.replaceAll("[^-._a-zA-Z0-9]+", "-").replaceAll("-+", "-");
```

Comment:
This is an  part of algorightm from namespace name normalization. Do we need to apply length and `replaceAll("^-|-$", "")` rule?
``` 
 @VisibleForTesting
  String normalizeNamespaceName(String namespaceName) {
    namespaceName =
        namespaceName
            .replaceAll("[^-a-zA-Z0-9]", "-") // replace invalid chars with '-'
            .replaceAll("-+", "-") // replace multiple '-' with single ones
            .replaceAll("^-|-$", ""); // trim dashes at beginning/end of the string

    return namespaceName.substring(
        0,
        Math.min(
            namespaceName.length(),
            METADATA_NAME_MAX_LENGTH)); // limit length to METADATA_NAME_MAX_LENGTH
  }
```

Reply:
The k8s documentations only says this about the keys of data:
The keys of data and stringData must consist of alphanumeric characters, `-`, `_` or `.`.

So I tested it manually and it looks like we need the length restriction to 253 chars but not the dash trimming because I was able to create a secret with dashes at both beginning and end.
![Screenshot from 2021-09-27 12-21-37](https://user-images.githubusercontent.com/73115616/134892439-b69411b5-aade-43cf-8012-bda1a1cd3b2a.png)